### PR TITLE
Batch mesolve

### PR DESCRIPTION
**Pull request by abocquet**
Here I use the convention that the dissipators batching dimension is always outwards because it makes the reasoning much simpler for this code, but if there is a good reason to adopt another one, we can switch.

On top of #436

### CODE REVIEW

File: dynamiqs/mesolve/mesolve.py [hunk 1]
```diff
@@ -27,6 +28,39 @@ def mesolve(
...
+        options.cartesian_batching,
+        args,
+        out_axes,
+        H.ndim > 2,
+        jump_ops.ndim > 3,
+        psi0.ndim > 2,
+    )
+
+    # === apply vectorized function
+    return f(H, jump_ops, psi0, tsave, exp_ops, solver, gradient, options)
```
Comment (gautierronan) [inline dynamiqs/mesolve/mesolve.py:51]:
```suggestion
    # compute in_axes to vectorize over
    batch_axes = (H.ndim > 2, jump_ops.ndim > 3, psi0.ndim > 2, None, None, None, None, None)
    in_axes = compute_in_axes(batch_axes, options.cartesian_batching)
    
    # the result is vectorized over ysave and Esave
    out_axes = Result(None, None, None, None, 0, 0)

    batched_mesolve = jax.vmap(_mesolve, in_axes=in_axes, out_axes=out_axes)

    # === apply vectorized function
    return batched_mesolve(H, jump_ops, psi0, tsave, exp_ops, solver, gradient, options)
```

File: dynamiqs/_utils.py [hunk 2]
```diff
@@ -41,3 +42,27 @@ def bexpect(O: Array, x: Array) -> Array:
     if isket(x):
         return jnp.einsum('ij,bjk,kl->b', dag(x), O, x)  # <x|O|x>
     return jnp.einsum('bij,ji->b', O, x)  # tr(Ox)
+
+
+def compute_batching(f, cartesian_batching, other_in_axes, out_axes, *should_batch):
```
Comment (gautierronan) [inline dynamiqs/_utils.py:47]:
```suggestion
def compute_in_axes(batch_axes, cartesian_batching):
```
(and corresponding function modifications so as to return the `in_axes` argument)

**Review by gautierronan — CHANGES_REQUESTED** [review]
Nice! I like the overall idea of the PR, but I'd suggest the changes below to remedy two issues that I find in the current form:
 - it's lacking readability (mesolve is the first function people will look at when reading source code: it should be very easy to follow IMO),
 - `compute_batching` does not support batching over arguments that are not within the first function arguments (useful for `smesolve`).

**Comment by abocquet** [comment]
I understand your concern but 
- the code proposal does not work as is since we need a for loop to vmap properly. If we chose to split the code according to your comment the loop would be in the mesolve code. The complexity does not dise appear, it is moved outside of `compute_batching` 
- one does not need to understand the batching mechanism during the first read. The goal of `compute_batching` is to abstract this mechanism away. This way of splitting concerns allows for understanding mesolve aside for batching and come back later if this is the part of interest 
- regarding smesolve this is a problem we can address later and probably the `ntraj` dimension (if this is what you are thinking about), could be in the first arguments of `_smsesolve` which solves the problem

**Comment by abocquet** [comment]
I shot myself with git but the PR is still open, we can continue discussing it while I open a new PR

**Pull request closed without merging**